api/http/handler/edgestacks/edgestack_status_update.go

```diff
@@ -94,66 +122,43 @@ func (handler *Handler) updateEdgeStackStatus(tx dataservices.DataStoreTx, r *ht
 		return nil, httperror.Forbidden("Permission denied to access environment", err)
 	}
 
-	var stack *portainer.EdgeStack
+	status := *payload.Status
+
+	log.Debug().
+		Int("stackID", int(stackID)).
+		Int("status", int(status)).
+		Msg("Updating stack status")
 
 	if featureflags.IsEnabled(portainer.FeatureNoTx) {
-		err = tx.EdgeStack().UpdateEdgeStackFunc(portainer.EdgeStackID(stackID), func(edgeStack *portainer.EdgeStack) {
-			details := edgeStack.Status[payload.EndpointID].Details
-			details.Pending = false
-
-			switch *payload.Status {
-			case portainer.EdgeStackStatusOk:
-				details.Ok = true
-			case portainer.EdgeStackStatusError:
-				details.Error = true
-			case portainer.EdgeStackStatusAcknowledged:
-				details.Acknowledged = true
-			case portainer.EdgeStackStatusRemove:
-				details.Remove = true
-			case portainer.EdgeStackStatusImagesPulled:
-				details.ImagesPulled = true
-			}
-
-			edgeStack.Status[payload.EndpointID] = portainer.EdgeStackStatus{
-				Details:    details,
-				Error:      payload.Error,
-				EndpointID: payload.EndpointID,
-			}
+		err = tx.EdgeStack().UpdateEdgeStackFunc(stackID, func(edgeStack *portainer.EdgeStack) {
+
+			environmentStatus.Status = append(environmentStatus.Status, portainer.EdgeStackDeploymentStatus{
```

Comment:
This needs to mutate `edgeStack`, without using external old copies as a reference, otherwise it introduces a last-write-wins situation.